File: DSA/binary_tree/bintree.c
// implement a binary tree

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

typedef struct node {
    int data;
    struct node *left;
    struct node *right;
} node;

node *createNode(int data) {
    node *newNode = (node *)malloc(sizeof(node));
    newNode->data = data;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

node *insertNode(node *root, int data) {
    if (root == NULL) {
        root = createNode(data);
    } else if (data <= root->data) {
        root->left = insertNode(root->left, data);
    } else {
        root->right = insertNode(root->right, data);
    }
    return root;
}

int searchNode(node *root, int data) {
    if (root == NULL) {
        return 0;
    } else if (root->data == data) {
        return 1;
    } else if (data <= root->data) {
        return searchNode(root->left, data);
    } else {
        return searchNode(root->right, data);
    }
}

void Inorder(node *root) {
    if (root == NULL) {
        return;
    }

    Inorder(root->left);
    printf("%d->", root->data);
    Inorder(root->right);
}

void Preorder(node *root){
    if(root == NULL){
        return;
    }
    printf("%d->", root->data);
    Preorder(root->left);
    Preorder(root->right);

}

// populate a tree with random numbers

node *populate_tree(node *root, int n, int max) {
    srand(time(NULL));
    for (int i = 0; i < n; i++) {
        int data = rand() % max;
        root = insertNode(root, data);
    }

    return root;
}


int main() {
    node *root = NULL;

    root = populate_tree(root, 10, 100);

    insertNode(root, 10);
    

    Inorder(root);
}

File: DSA/binary_tree/bintree.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None
        
class BinaryTree:
    def __init__(self, root):
        self.root = Node(root)
        
    def print_tree(self, traversal_type):
        if traversal_type == "preorder":
            return self.preorder_print(self.root, "")
        elif traversal_type == "inorder":
            return self.inorder_print(self.root, "")
        elif traversal_type == "postorder":
            return self.postorder_print(self.root, "")
        else:
            print("Traversal type " + str(traversal_type) + " is not supported.")
            return False
        
    def preorder_print(self, start, traversal):
        """Root->Left->Right"""
        if start:
            traversal += (str(start.value) + "->")
            traversal = self.preorder_print(start.left, traversal)
            traversal = self.preorder_print(start.right, traversal)
        return traversal
    
    def inorder_print(self, start, traversal):
        """Left->Root->Right"""
        if start:
            traversal = self.inorder_print(start.left, traversal)
            traversal += (str(start.value) + "->")
            traversal = self.inorder_print(start.right, traversal)
        return traversal
    
    def postorder_print(self, start, traversal):
        """Left->Right->Root"""
        if start:
            traversal = self.postorder_print(start.left, traversal)
            traversal = self.postorder_print(start.right, traversal)
            traversal += (str(start.value) + "->")
        return traversal

    def insert(self, value):
        self.insert_helper(self.root, value)

    def insert_helper(self, current_node, value):
        if value < current_node.value:
            if current_node.left:
                self.insert_helper(current_node.left, value)
            else:
                current_node.left = Node(value)
        else:
            if current_node.right:
                self.insert_helper(current_node.right, value)
            else:
                current_node.right = Node(value)

    def delete(self, value):
        self.delete_helper(self.root, value)

    def delete_helper(self, current_node, value):
        if current_node:
            if value < current_node.value:
                current_node.left = self.delete_helper(current_node.left, value)
            elif value > current_node.value:
                current_node.right = self.delete_helper(current_node.right, value)
            else:
                if current_node.left is None and current_node.right is None:
                    current_node = None
                elif current_node.left is None:
                    current_node = current_node.right
                elif current_node.right is None:
                    current_node = current_node.left
                else:
                    temp_node = self.get_max(current_node.left)
                    current_node.value = temp_node.value
                    current_node.left = self.delete_helper(current_node.left, temp_node.value)
        return current_node

    def get_max(self, current_node):
        if current_node.right:
            return self.get_max(current_node.right)
        return current_node

# populate tree

tree = BinaryTree(1)

def populate_tree(tree, num_elems = 10, max_int = 100):
    from random import randint
    for _ in range(num_elems):
        current_elem = randint(0, max_int)
        tree.insert(current_elem)
    return tree

populate_tree(tree, 200**10, 100000)

# print tree

print(tree.print_tree("inorder"))


# reverse tree

def reverse_tree(tree):
    def reverse_helper(node):
        if node:
            node.left, node.right = node.right, node.left
            reverse_helper(node.left)
            reverse_helper(node.right)
    reverse_helper(tree.root)
    return tree

